cartridge/shop/managers.py

```diff
@@ -22,26 +22,20 @@ def from_request(self, request):
         cart.
         """
         cart_id = request.session.get("cart", None)
-        cart = None
-        if cart_id:
+        cart = self.current().filter(id=cart_id)
+        last_updated = now()
+        # Update timestamp and clear out old carts.
+        if cart_id and cart.update(last_updated=last_updated):
+            self.expired().delete()
+        else:
             try:
-                cart = self.current().get(id=cart_id)
-            except self.model.DoesNotExist:
                 request.session["cart"] = None
-            else:
-                # Update timestamp and clear out old carts.
-                cart.last_updated = now()
-                cart.save()
-                self.expired().delete()
-        if not cart:
-            # Forget what checkout step we were up to.
-            try:
+                # Forget what checkout step we were up to.
                 del request.session["order"]["step"]
                 request.session.modified = True
             except KeyError:
                 pass
-            cart = self.model(last_updated=now())
-        return cart
+        return self.model(id=cart_id, last_updated=last_updated)
```

Comment:
Yep, I'd contemplated that as well. However, I think the write every request is the bigger concern. I'd tried to contemplate a way to tie the cart expiration more to the session rather than writing to the cart table every time, but didn't really come up with anything yet.

Since carts aren't visible in admin (although they could be useful/fun there), it would be possible to do some clever DB routing on my project and simply store carts in a local database, since a user should always end up on the same web server. And in the off chance they don't, you could have rudimentary failover in the DB router to go check the real database in case they ended up over there. But that's just me musing aloud.

I'm going to go ahead and make the changes discussed here in the branch for simplicity.